Upgrade ignores "version" property of "depends"

**CKAN Version:** 
1.24.0

**KSP Version:** 
1.3.1

**Operating System:** 
Windows 10

**Have you made any manual changes to your GameData folder (i.e., not via CKAN)?** 
no

Problem
-------

**What steps did you take in CKAN?**
On a clean installation of KSP 1.3.1 open ckan.exe, select the mod StockalikeSolarSystem and install.

Notice that this mod requires a specific version of other mods to be installed.

Once the installation is complete, click on "Add Available Updates". Then Apply,

**What happened?**

some of the mods that were installed at a specific version to fulfill StockalikeSolarSystem requirements have been updated to the latest version, breaking compatibility with SASS and potentially damaging saves if the player ignores the warnings

**What did you expect to happen?**
some kind of warning, or red flag signalling that the update was not recommended.


Came up during discussion of #2319.

OK, better test...

1. Installed SASS with the current normal metadata
2. Added the above `"conflicts"` block to SASS in `installed_modules`

```
$ ckan upgrade --all
2281 [1] ERROR CKAN.RegistryManager (null) - Loaded registry with inconsistencies:

The following inconsistencies were found:
StockalikeSolarSystem conflicts with Kopernicus.

Upgrading modules...


Unhandled Exception:
The following inconsistencies were found:
StockalikeSolarSystem conflicts with Kopernicus.  at CKAN.SanityChecker.EnforceConsistency (System.Collections.Generic.IEnumerable`1[T] modules, System.Collections.Generic.IEnumerable`1[T] dlls) [0x00017] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.Registry.CheckSanity () [0x00035] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.RegistryManager.Save (System.Boolean enforce_consistency) [0x00018] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.ModuleInstaller.AddRemove (System.Collections.Generic.IEnumerable`1[T] add, System.Collections.Generic.IEnumerable`1[T] remove, System.Boolean enforceConsistency) [0x0005f] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.ModuleInstaller.Upgrade (System.Collections.Generic.IEnumerable`1[T] modules, CKAN.IDownloader netAsyncDownloader, System.Boolean enforceConsistency) [0x00130] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.CmdLine.Upgrade.RunCommand (CKAN.KSP ksp, System.Object raw_options) [0x002be] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.CmdLine.MainClass.RunSimpleAction (CKAN.CmdLine.Options cmdline, CKAN.CmdLine.CommonOptions options, System.String[] args, CKAN.IUser user, CKAN.KSPManager manager) [0x0040a] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.CmdLine.MainClass.Execute (CKAN.KSPManager manager, CKAN.CmdLine.CommonOptions opts, System.String[] args) [0x0015b] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.CmdLine.MainClass.Main (System.String[] args) [0x000a8] in <cf1172f4644841da8b9b83c88c51f181>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: The following inconsistencies were found:
StockalikeSolarSystem conflicts with Kopernicus.  at CKAN.SanityChecker.EnforceConsistency (System.Collections.Generic.IEnumerable`1[T] modules, System.Collections.Generic.IEnumerable`1[T] dlls) [0x00017] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.Registry.CheckSanity () [0x00035] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.RegistryManager.Save (System.Boolean enforce_consistency) [0x00018] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.ModuleInstaller.AddRemove (System.Collections.Generic.IEnumerable`1[T] add, System.Collections.Generic.IEnumerable`1[T] remove, System.Boolean enforceConsistency) [0x0005f] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.ModuleInstaller.Upgrade (System.Collections.Generic.IEnumerable`1[T] modules, CKAN.IDownloader netAsyncDownloader, System.Boolean enforceConsistency) [0x00130] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.CmdLine.Upgrade.RunCommand (CKAN.KSP ksp, System.Object raw_options) [0x002be] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.CmdLine.MainClass.RunSimpleAction (CKAN.CmdLine.Options cmdline, CKAN.CmdLine.CommonOptions options, System.String[] args, CKAN.IUser user, CKAN.KSPManager manager) [0x0040a] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.CmdLine.MainClass.Execute (CKAN.KSPManager manager, CKAN.CmdLine.CommonOptions opts, System.String[] args) [0x0015b] in <cf1172f4644841da8b9b83c88c51f181>:0 
  at CKAN.CmdLine.MainClass.Main (System.String[] args) [0x000a8] in <cf1172f4644841da8b9b83c88c51f181>:0 
```

So once again we see that conflict checking ignores versions, which makes CKAN think that this install is inconsistent even before the upgrade, even though it shouldn't be.

--- a/Cmdline/Action/Upgrade.cs
+++ b/Cmdline/Action/Upgrade.cs
@@ -128,6 +128,11 @@ namespace CKAN.CmdLine
                 User.RaiseMessage("Module {0} not found", kraken.module);
                 return Exit.ERROR;
             }
+            catch (InconsistentKraken kraken)
+            {
+                User.RaiseMessage(kraken.ToString());
+                return Exit.ERROR;
+            }
             User.RaiseMessage("\r\nDone!\r\n");
 
             return Exit.OK;

--- a/Core/Registry/Registry.cs
+++ b/Core/Registry/Registry.cs
@@ -1001,9 +1001,8 @@ namespace CKAN
                 log.DebugFormat("Started with {0}, removing {1}, and keeping {2}; our dlls are {3}", string.Join(", ", orig_installed), string.Join(", ", modules_to_remove), string.Join(", ", hypothetical), string.Join(", ", dlls));
 
                 // Find what would break with this configuration.
-                // The Values.SelectMany() flattens our list of broken mods.
-                var broken = new HashSet<string>(SanityChecker.FindUnmetDependencies(hypothetical, dlls)
-                    .Values.SelectMany(x => x).Select(x => x.identifier));
+                var broken = SanityChecker.FindUnsatisfiedDepends(hypothetical, dlls.ToHashSet())
+                    .Select(x => x.Key.identifier).ToHashSet();
 
                 // If nothing else would break, it's just the list of modules we're removing.
                 HashSet<string> to_remove = new HashSet<string>(modules_to_remove);

--- a/Core/Relationships/SanityChecker.cs
+++ b/Core/Relationships/SanityChecker.cs
@@ -9,7 +9,6 @@ namespace CKAN
     /// </summary>
     public static class SanityChecker
     {
-
         private static readonly ILog log = LogManager.GetLogger(typeof(SanityChecker));
 
         /// <summary>
@@ -26,57 +25,17 @@ namespace CKAN
                 return errors;
             }
 
-            Dictionary<string, List<string>> providers = ModulesToProvides(modules, dlls);
-
-            foreach (KeyValuePair<string,List<CkanModule>> entry in FindUnmetDependencies(modules, dlls))
+            foreach (var kvp in FindUnsatisfiedDepends(modules.ToList(), dlls?.ToHashSet()))
             {
-                foreach (CkanModule unhappy_mod in entry.Value)
-                {
-                    // This error can fire if a dependency
-                    // IS listed in the index AND IS available for our version,
-                    // but not installed.
-                    // This happens when `dlls` is Registry.installed_dlls.Keys,
-                    // as in Registry.GetSanityErrors.
-                    errors.Add(string.Format(
-                        "{0} has an unmet dependency: {1} is not installed",
-                        unhappy_mod.identifier,
-                        entry.Key
-                    ));
-                }
+                errors.Add($"{kvp.Key} has an unsatisfied dependency: {kvp.Value} is not installed");
             }
 
             // Conflicts are more difficult. Mods are allowed to conflict with themselves.
             // So we walk all our mod conflicts, find what (if anything) provide those
             // conflicts, and return false if it's not the module we're examining.
-
-            // TODO: This doesn't examine versions. We should!
-            // TODO: It would be great to factor this into its own function, too.
-
-            foreach (CkanModule mod in modules)
+            foreach (var kvp in FindConflicting(modules.ToList(), dlls?.ToHashSet()))
             {
-                // If our mod doesn't conflict with anything, skip it.
-                if (mod.conflicts == null)
-                {
-                    continue;
-                }
-
-                foreach (var conflict in mod.conflicts)
-                {
-                    // If nothing conflicts with us, skip.
-                    if (! providers.ContainsKey(conflict.name))
-                    {
-                        continue;
-                    }
-
-                    // If something does conflict with us, and it's not ourselves, that's a fail.
-                    foreach (string provider in providers[conflict.name])
-                    {
-                        if (provider != mod.identifier)
-                        {
-                            errors.Add(string.Format("{0} conflicts with {1}.", mod.identifier, provider));
-                        }
-                    }
-                }
+                errors.Add($"{kvp.Key} conflicts with {kvp.Value}");
             }
 
             // Return whatever we've found, which could be empty.
@@ -97,7 +56,7 @@ namespace CKAN
                 throw new InconsistentKraken(errors);
             }
         }
-    
+
         /// <summary>
         /// Returns true if the mods supplied can co-exist. This checks depends/pre-depends/conflicts only.
         /// </summary>
@@ -107,93 +66,64 @@ namespace CKAN
         }
 
         /// <summary>
-        /// Maps a list of modules and dlls to a dictionary of what's provided, and a list of
-        /// identifiers that supply each.
+        /// Find unsatisfied dependencies among the given modules and DLLs.
         /// </summary>
-        // 
-        // Eg: {
-        //          LifeSupport => [ "TACLS", "Snacks" ]
-        //          DogeCoinFlag => [ "DogeCoinFlag" ]
-        // }
-        public static Dictionary<string, List<string>> ModulesToProvides(IEnumerable<CkanModule> modules, IEnumerable<string> dlls = null)
+        /// <param name="modules">List of modules to check</param>
+        /// <param name="dlls">List of DLLs that can also count toward relationships</param>
+        /// <returns>
+        /// List of dependencies that aren't satisfied represented as pairs.
+        /// Each Key is the depending module, and each Value is the relationship.
+        /// </returns>
+        public static List<KeyValuePair<CkanModule, RelationshipDescriptor>>
+            FindUnsatisfiedDepends(IEnumerable<CkanModule> modules, HashSet<string> dlls)
         {
-            var providers = new Dictionary<string, List<string>>();
-
-            if (dlls == null)
+            var unsat = new List<KeyValuePair<CkanModule, RelationshipDescriptor>>();
+            if (modules != null)
             {
-                dlls = new List<string>();
-            }
-
-            foreach (CkanModule mod in modules)
-            {
-                foreach (string provides in mod.ProvidesList)
+                foreach (CkanModule m in modules.Where(m => m.depends != null))
                 {
-                    log.DebugFormat("{0} provides {1}", mod, provides);
-                    providers[provides] = providers.ContainsKey(provides) ? providers[provides] : new List<string>();
-                    providers[provides].Add(mod.identifier);
-                }
-            }
-
-            // Add in our DLLs as things we know exist.
-            foreach (string dll in dlls)
-            {
-                if (! providers.ContainsKey(dll))
-                {
-                    providers[dll] = new List<string>();
-                }
-                providers[dll].Add(dll);
-            }
-
-            return providers;
-        }
-
-        /// <summary>
-        /// Given a list of modules and optional dlls, returns a dictionary of dependencies which are still not met.
-        /// The dictionary keys are the un-met depdendencies, the values are the modules requesting them.
-        /// </summary>
-        public static Dictionary<string,List<CkanModule>> FindUnmetDependencies(IEnumerable<CkanModule> modules, IEnumerable<string> dlls = null)
-        {
-            return FindUnmetDependencies(modules, ModulesToProvides(modules, dlls));
-        }
-
-        /// <summary>
-        /// Given a list of modules, and a dictionary of providers, returns a dictionary of depdendencies which have not been met.
-        /// </summary>
-        internal static Dictionary<string,List<CkanModule>> FindUnmetDependencies(IEnumerable<CkanModule> modules, Dictionary<string,List<string>> provided)
-        {
-            return FindUnmetDependencies(modules, new HashSet<string> (provided.Keys));
-        }
-
-        /// <summary>
-        /// Given a list of modules, and a set of providers, returns a dictionary of dependencies which have not been met.
-        /// </summary>
-        internal static Dictionary<string,List<CkanModule>> FindUnmetDependencies(IEnumerable<CkanModule> modules, HashSet<string> provided)
-        {
-            var unmet = new Dictionary<string,List<CkanModule>> ();
-
-            // TODO: This doesn't examine versions, it should!
-
-            foreach (CkanModule mod in modules)
-            {
-                // If this module has no dependencies, we're done.
-                if (mod.depends == null)
-                {
-                    continue;
-                }
-
-                // If it does have dependencies, but we can't find anything that provides them,
-                // add them to our unmet list.
-                foreach (RelationshipDescriptor dep in mod.depends.Where(dep => ! provided.Contains(dep.name)))
-                {
-                    if (!unmet.ContainsKey(dep.name))
+                    foreach (RelationshipDescriptor dep in m.depends)
                     {
-                        unmet[dep.name] = new List<CkanModule>();
+                        if (!dep.MatchesAny(modules, dlls))
+                        {
+                            unsat.Add(new KeyValuePair<CkanModule, RelationshipDescriptor>(m, dep));
+                        }
                     }
-                    unmet[dep.name].Add(mod); // mod needs dep.name, but doesn't have it.
                 }
             }
-
-            return unmet;
+            return unsat;
         }
+
+        /// <summary>
+        /// Find conflicts among the given modules and DLLs.
+        /// </summary>
+        /// <param name="modules">List of modules to check</param>
+        /// <param name="dlls">List of DLLs that can also count toward relationships</param>
+        /// <returns>
+        /// List of conflicts represented as pairs.
+        /// Each Key is the depending module, and each Value is the relationship.
+        /// </returns>
+        public static List<KeyValuePair<CkanModule, RelationshipDescriptor>>
+            FindConflicting(IEnumerable<CkanModule> modules, HashSet<string> dlls)
+        {
+            var confl = new List<KeyValuePair<CkanModule, RelationshipDescriptor>>();
+            if (modules != null)
+            {
+                foreach (CkanModule m in modules.Where(m => m.conflicts != null))
+                {
+                    // Remove self from the list, so we're only comparing to OTHER modules
+                    var others = modules.Where(other => other != m);
+                    foreach (RelationshipDescriptor dep in m.conflicts)
+                    {
+                        if (dep.MatchesAny(others, dlls))
+                        {
+                            confl.Add(new KeyValuePair<CkanModule, RelationshipDescriptor>(m, dep));
+                        }
+                    }
+                }
+            }
+            return confl;
+        }
+
     }
 }

--- a/Core/Types/CkanModule.cs
+++ b/Core/Types/CkanModule.cs
@@ -53,6 +53,44 @@ namespace CKAN
         }
 
         /// <summary>
+        /// Check whether any of the modules in a given list match this descriptor.
+        /// NOTE: Only proper modules can be checked for versions!
+        ///       DLLs match all versions, as do "provides" clauses.
+        /// </summary>
+        /// <param name="modules">Sequence of modules to consider</param>
+        /// <param name="dlls">Sequence of DLLs to consider</param>
+        /// <returns>
+        /// true if any of the modules match this descriptor, false otherwise.
+        /// </returns>
+        public bool MatchesAny(IEnumerable<CkanModule> modules, HashSet<string> dlls)
+        {
+            // DLLs are considered to match any version
+            if (dlls != null && dlls.Contains(name))
+            {
+                return true;
+            }
+            if (modules != null)
+            {
+                // See if anyone else "provides" the target name
+                // Note that versions can't be checked for "provides" clauses
+                if (modules.Any(m =>
+                    m.identifier != name && m.provides != null && m.provides.Contains(name)))
+                {
+                    return true;
+                }
+                // See if the real thing is there
+                foreach (CkanModule m in modules.Where(m => m.identifier == name))
+                {
+                    if (version_within_bounds(m.version))
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        /// <summary>
         /// A user friendly message for what versions satisfies this descriptor.
         /// </summary>
         [JsonIgnore]
@@ -66,6 +104,27 @@ namespace CKAN
                     min_version != null ? min_version.ToString() : "any version",
                     max_version != null ? max_version.ToString() : "any version");
             }
+        }
+
+        /// <summary>
+        /// Generate a user readable description of the relationship
+        /// </summary>
+        /// <returns>
+        /// Depending on the version properties, one of:
+        /// name
+        /// name version
+        /// name min_version -- max_version
+        /// name min_version or later
+        /// name max_version or earlier
+        /// </returns>
+        public override string ToString()
+        {
+            return
+                  version     != null                        ? $"{name} {version}"
+                : min_version != null && max_version != null ? $"{name} {min_version} -- {max_version}"
+                : min_version != null                        ? $"{name} {min_version} or later"
+                : max_version != null                        ? $"{name} {max_version} or earlier"
+                : name;
         }
 
     }

--- a/Core/Types/Kraken.cs
+++ b/Core/Types/Kraken.cs
@@ -1,4 +1,5 @@
 using System;
+using System.Linq;
 using System.Net;
 using System.Text;
 using System.Collections.Generic;
@@ -154,7 +155,7 @@ namespace CKAN
         internal static string FormatMessage(string requested, List<CkanModule> modules)
         {
             string oops = string.Format("Too many mods provide {0}:\r\n", requested);
-            return oops + String.Join("\r\n* ", modules);
+            return oops + String.Join("\r\n", modules.Select(m => $"* {m}"));
         }
     }
 
@@ -170,8 +171,7 @@ namespace CKAN
         {
             get
             {
-                const string message = "The following inconsistencies were found:\r\n";
-                return message + String.Join("\r\n * ", inconsistencies);
+                return header + String.Join("\r\n", inconsistencies.Select(msg => $"* {msg}"));
             }
         }
 
@@ -189,8 +189,10 @@ namespace CKAN
 
         public override string ToString()
         {
-            return InconsistenciesPretty + StackTrace;
+            return InconsistenciesPretty + "\r\n\r\n" + StackTrace;
         }
+
+        private const string header = "The following inconsistencies were found:\r\n";
     }
 
     /// <summary>

--- a/Tests/Core/Relationships/SanityChecker.cs
+++ b/Tests/Core/Relationships/SanityChecker.cs
@@ -74,17 +74,17 @@ namespace Tests.Core.Relationships
 
             // This would actually be a terrible thing for users to have, but it tests the
             // relationship we want.
-            mods.Add(registry.LatestAvailable("CustomBiomesKerbal",null));
+            mods.Add(registry.LatestAvailable("CustomBiomesKerbal", null));
             Assert.IsTrue(CKAN.SanityChecker.IsConsistent(mods, dlls), "CustomBiomes DLL, with config added");
 
-            mods.Add(registry.LatestAvailable("CustomBiomesRSS",null));
+            mods.Add(registry.LatestAvailable("CustomBiomesRSS", null));
             Assert.IsFalse(CKAN.SanityChecker.IsConsistent(mods, dlls), "CustomBiomes with conflicting data");
         }
 
         [Test]
         public void ConflictWithDll()
         {
-            var mods = new List<CkanModule> { registry.LatestAvailable("SRL",null) };
+            var mods = new List<CkanModule> { registry.LatestAvailable("SRL", null) };
             var dlls = new List<string> { "QuickRevert" };
 
             Assert.IsTrue(CKAN.SanityChecker.IsConsistent(mods), "SRL can be installed by itself");
@@ -92,43 +92,33 @@ namespace Tests.Core.Relationships
         }
 
         [Test]
-        public void ModulesToProvides()
-        {
-            var mods = new List<CkanModule>
-            {
-                registry.LatestAvailable("CustomBiomes",null),
-                registry.LatestAvailable("CustomBiomesKerbal",null),
-                registry.LatestAvailable("DogeCoinFlag",null)
-            };
-
-            var provides = CKAN.SanityChecker.ModulesToProvides(mods);
-            Assert.Contains("CustomBiomes", provides.Keys);
-            Assert.Contains("CustomBiomesData", provides.Keys);
-            Assert.Contains("CustomBiomesKerbal", provides.Keys);
-            Assert.Contains("DogeCoinFlag", provides.Keys);
-            Assert.AreEqual(4, provides.Keys.Count);
-        }
-
-        [Test]
-        public void FindUnmetDependencies()
+        public void FindUnsatisfiedDepends()
         {
             var mods = new List<CkanModule>();
-            var dlls = Enumerable.Empty<string>();
-            Assert.IsEmpty(CKAN.SanityChecker.FindUnmetDependencies(mods, dlls), "Empty list");
+            var dlls = Enumerable.Empty<string>().ToHashSet();
+            Assert.IsEmpty(CKAN.SanityChecker.FindUnsatisfiedDepends(mods, dlls), "Empty list");
 
-            mods.Add(registry.LatestAvailable("DogeCoinFlag",null));
-            Assert.IsEmpty(CKAN.SanityChecker.FindUnmetDependencies(mods, dlls), "DogeCoinFlag");
+            mods.Add(registry.LatestAvailable("DogeCoinFlag", null));
+            Assert.IsEmpty(CKAN.SanityChecker.FindUnsatisfiedDepends(mods, dlls), "DogeCoinFlag");
 
-            mods.Add(registry.LatestAvailable("CustomBiomes",null));
-            Assert.Contains("CustomBiomesData", CKAN.SanityChecker.FindUnmetDependencies(mods, dlls).Keys, "Missing CustomBiomesData");
+            mods.Add(registry.LatestAvailable("CustomBiomes", null));
+            Assert.Contains(
+                "CustomBiomesData",
+                CKAN.SanityChecker.FindUnsatisfiedDepends(mods, dlls).Select(kvp => kvp.Value.name).ToList(),
+                "Missing CustomBiomesData"
+            );
 
-            mods.Add(registry.LatestAvailable("CustomBiomesKerbal",null));
-            Assert.IsEmpty(CKAN.SanityChecker.FindUnmetDependencies(mods, dlls), "CBD+CBK");
+            mods.Add(registry.LatestAvailable("CustomBiomesKerbal", null));
+            Assert.IsEmpty(CKAN.SanityChecker.FindUnsatisfiedDepends(mods, dlls), "CBD+CBK");
 
             mods.RemoveAll(x => x.identifier == "CustomBiomes");
             Assert.AreEqual(2, mods.Count, "Checking removed CustomBiomes");
 
-            Assert.Contains("CustomBiomes", CKAN.SanityChecker.FindUnmetDependencies(mods, dlls).Keys, "Missing CustomBiomes");
+            Assert.Contains(
+                "CustomBiomes",
+                CKAN.SanityChecker.FindUnsatisfiedDepends(mods, dlls).Select(kvp => kvp.Value.name).ToList(),
+                "Missing CustomBiomes"
+            );
         }
 
         [Test]

